src/GCHP_GridComp/GCHPctmEnv_GridComp/GCHPctmEnv_GridCompMod.F90

```diff
@@ -660,53 +634,109 @@ subroutine Run ( GC, IMPORT, EXPORT, CLOCK, RC )
       DryPLE0r8(:,:,:) = DryPLE0r8(:,:,LM:0:-1)
       DryPLE1r8(:,:,:) = DryPLE1r8(:,:,LM:0:-1)
       PLE0r8   (:,:,:) = PLE0r8   (:,:,LM:0:-1)
       PLE1r8   (:,:,:) = PLE1r8   (:,:,LM:0:-1)
-      UC       (:,:,:) =  UC      (:,:,LM:1:-1)
-      VC       (:,:,:) =  VC      (:,:,LM:1:-1)
-
-      DEALLOCATE( AP, BP )
+      SPHU0r8  (:,:,:) = 1.0d0*SPHU0(:,:,LM:1:-1)
+      
+      ! Deallocate temporaries
+      DEALLOCATE( AP, BP, SPHU0, SPHU1 )
 
-      call MAPL_GetPointer ( EXPORT, MFXr8, 'MFXr8', RC=STATUS )
+      !! Calculate MF[XY] and C[XY] exports
+      ! Exports
+      call MAPL_GetPointer ( EXPORT, MFX, 'MFX', RC=STATUS )
       _VERIFY(STATUS)
-      call MAPL_GetPointer ( EXPORT, MFYr8, 'MFYr8', RC=STATUS )
+      call MAPL_GetPointer ( EXPORT, MFY, 'MFY', RC=STATUS )
       _VERIFY(STATUS)
-      call MAPL_GetPointer ( EXPORT,  CXr8,  'CXr8', RC=STATUS )
+      call MAPL_GetPointer ( EXPORT,  CX,  'CX', RC=STATUS )
       _VERIFY(STATUS)
-      call MAPL_GetPointer ( EXPORT,  CYr8,  'CYr8', RC=STATUS )
+      call MAPL_GetPointer ( EXPORT,  CY,  'CY', RC=STATUS )
       _VERIFY(STATUS)
+      if (.not. import_mass_flux_from_extdata) then
+         !! Calculate MFX, MFY, CX, and CY exports (with top-down indexing)
 
-      ! Compute the courant numbers and mass fluxes
-      !--------------------------------------------
-      ALLOCATE( UCr8(is:ie,js:je,lm),   STAT=STATUS); _VERIFY(STATUS)
-      ALLOCATE( VCr8(is:ie,js:je,lm),   STAT=STATUS); _VERIFY(STATUS)
-      ALLOCATE(PLEr8(is:ie,js:je,lm+1), STAT=STATUS); _VERIFY(STATUS)
+         ! Imports
+         call MAPL_GetPointer ( IMPORT,      UA,    'UA',  RC=STATUS )
+         _VERIFY(STATUS)
+         call MAPL_GetPointer ( IMPORT,      VA,    'VA',  RC=STATUS )
+         _VERIFY(STATUS)
 
-      UCr8  = 1.00d0*(UC)
-      VCr8  = 1.00d0*(VC)
+         ! Temporaries
+         ALLOCATE(UC(is:ie,js:je,lm), STAT=STATUS); 
+         _VERIFY(STATUS)
+         ALLOCATE(VC(is:ie,js:je,lm), STAT=STATUS); 
+         _VERIFY(STATUS)
+         ALLOCATE(UCr8(is:ie,js:je,lm), STAT=STATUS); 
+         _VERIFY(STATUS)
+         ALLOCATE(VCr8(is:ie,js:je,lm), STAT=STATUS); 
+         _VERIFY(STATUS)
+         ALLOCATE(PLEr8(is:ie,js:je,lm+1), STAT=STATUS); 
+         _VERIFY(STATUS)
 
-      ! Use dry pressure at the start of the timestep to calculate mass
-      ! fluxes. GMAO method uses mid-step UC, VC and PLE?
-      PLEr8 = 1.00d0*(DryPLE0r8)
+         ! Prepare inputs to fv_computeMassFluxes
+
+         ! Copy UA,VA to UC,VC while handling top-down vs. bottom-up meteorological data
+         if (meteorology_vertical_index_is_top_down) then
+            UC(:,:,:) = UA(:,:,LM:1:-1)
+            VC(:,:,:) = VA(:,:,LM:1:-1)
+         else
+            UC(:,:,:) = UA(:,:,:)
+            VC(:,:,:) = VA(:,:,:)
+         end if
+         ! Restagger winds (A-grid to C-grid)
+         call A2D2C(U=UC, V=VC, npz=lm, getC=.true.)
```

Comment:
I wonder if A2D2C needs the vertical flipping. If it applies the same computations to levels with no exceptions then it seems the previous vertical flip is not necessary, and the flip later on could be consolidated.

Reply:
Added in #181 (merged into this PR)